samza-core/src/main/java/org/apache/samza/operators/impl/WatermarkStates.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.operators.impl;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.system.WatermarkMessage;
+
+import static org.apache.samza.operators.functions.WatermarkFunction.WATERMARK_NOT_EXIST;
+
+
+/**
+ * This class manages the state of the watermarks in a task. Internally it keeps track of the latest
+ * watermark timestamp from each upstream task, and use the min as the consolidated watermark time.
+ */
+class WatermarkStates {
+
+  private final static class WatermarkState {
+    private int expectedTotal = Integer.MAX_VALUE;
+    private final Map<String, Long> timestamps = new HashMap<>();
+    private long watermarkTime = WATERMARK_NOT_EXIST;
+
+    WatermarkState(int expectedTotal) {
+      this.expectedTotal = expectedTotal;
+    }
+
+    boolean update(long timestamp, String taskName) {
+      final long preWatermarkTime = watermarkTime;
+      if (taskName != null) {
+        timestamps.put(taskName, timestamp);
+      }
+
+      if (timestamps.size() == expectedTotal) {
+        Optional<Long> min = timestamps.values().stream().min(Long::compare);
+        watermarkTime = min.orElse(timestamp);
+      }
+
+      return preWatermarkTime != watermarkTime;
+    }
+
+    long getWatermarkTime() {
+      return watermarkTime;
+    }
+  }
+
+  private final Map<SystemStreamPartition, WatermarkState> watermarkStates;
```

Comment:
Same comment as above re. thread-safety.

Reply:
Added the doc on the WatermarkStates class for thread-safety.